feat: support Neovim versions older than 0.10

Andrew15-5 opened the pull request:
I noticed that it uses `vim.uv`, which I happen to know is the new thing in 0.10. I also know that some plugins have a hack for it. So I looked it up and copied it over.

But I actually haven't tested it in older versions. If it's the only breaking change that needs patching then it should just work.

Andrew15-5 committed:
feat: support Neovim versions older than 0.10

PartyWumpus commented:
Did you accidentally commit the wrong thing? All this adds is a linevthat says `https://github.com/PartyWumpus/typst-concealer600f87c`?

Andrew15-5 commented:
Yep...

For some reason, opening this file gave me a treesitter error that permanently shows up (v0.10), so I thought that I might as well switch to the nixpkgs version. And I did. But now `dd` and `cc` don't work, as they freeze the editor entirely. And so this was added and removed during the cloning and commiting. No idea why this is still present and the thing that I added is not...

The `dd` bug has something to do with the treesitter plugin being enabled.

```
Error in decoration provider treesitter/highlighter.win:
Error executing lua: /usr/local/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 5:8. Invalid node type "string_content":
      (string_content) @injection.content)))
       ^
                                                                                                                                                                                                                     
stack traceback:
        [C]: in function '_ts_parse_query'
        /usr/local/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'
        /usr/local/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'
        /usr/local/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:128: in function 'new'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:511: in function 'add_child'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:395: in function '_add_injections'
        ...l/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:447: in function 'parse'
        ...al/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:397: in function <...al/share/nvim/runtime/lua/vim/treesitter/
```

So it feels like you have put something illegal in your file that both versions have different bugs only in this file (well, `dd` bug also happens in other files). disabling treesitter also removes this permanent error.

Andrew15-5 commented:
The problem with the file exists because of this:

```lua
  M._typst_query = vim.treesitter.query.parse("typst", [[
[
 (code
  [(_) (call item: (ident) @call_ident)] @code
 )
 (math)
] @block
]])
```

If I even remove a few lines of it, then the error stops showing. Still not sure why it happens though.

PartyWumpus commented:
This plugin cannot viably support lower than nvim 10, as it relies on:
- Inline virt_text extmarks
- Multi-line extmark ranges
- `nvim_buf_get_extmarks`'s "overlap" option
- `vim.list_contains` (although I could not use that)

All of which were added in nvim 10.

(and in the future I intend to the `url` option for extmarks but obviously that won't be required for the plugin to work so that could be disabled for <10)

Instead I'm just going to add a check for the neovim version and fail if it's too low.

Andrew15-5 commented:
Yeah, I guess adding a `vim.notify` `ERROR` is good.

Also don't forget to update the README.
